src/main/java/com/visenze/productsearch/ProductSearch.java

```diff
@@ -0,0 +1,213 @@
+         * @param config custom client http configurations
+         *
+         * @return this 'itself'
+         *
+         * @see ClientConfig
+         */
+        public Builder setClientConfig(ClientConfig config) {
+            this.config = config;
+            return this;
+        }
+    }
+
+    /**
+     * Constructor of ProductSearch made private to only allow it's Builder
+     * class to create it.
+     *
+     * @param appKey Unique app key that acts as authenticator for the client
+     * @param placementId Placement ID of the template that was chosen when
+     *                     creating apps on the dashboard
+     * @param endpoint Which endpoint to use for all ViHttpClient queries
+     * @param config Configuration to the behaviours of the ViHttpClient
+     */
+    private ProductSearch(String appKey, Integer placementId, String endpoint,
+                          ClientConfig config)
+    {
+        this.appKey       = appKey;
+        this.placementId  = placementId;
+        this.endpoint     = endpoint;
+        this.httpClient   = new ViSearchHttpClientImpl(this.endpoint, appKey, placementId.toString(), config);
+    }
+
+    /**
+     * Calls the POST method for the image search API
+     *
+     * @param params ImageSearchParam specific parameters to perform request
+     *
+     * @return ViSearchHttpResponse http response of search results
+     */
+    public ProductSearchResponse imageSearch(SearchByImageParam params) {
+        // test for image parameter validity
+        final File imageFile = params.getImage();
+        // attempt search using image file
+        if (imageFile != null) {
+            try {
+                return ProductSearchResponse.fromResponse(httpClient.postImage(DEFAULT_IMAGE_SEARCH_PATH, params.toMultimap(), new FileInputStream(imageFile), imageFile.getName()));
+            } catch (FileNotFoundException e) {
+                throw new InternalViSearchException(ResponseMessages.INVALID_IMAGE_OR_URL, e);
+            }
+        }
+        Multimap<String, String> paramMap = params.toMultimap();
+        paramMap.put("app_key", this.appKey);
+        paramMap.put("placement_id", this.placementId.toString());
+        // attempt using post for image url or image id
+        return ProductSearchResponse.fromResponse(httpClient.post(DEFAULT_IMAGE_SEARCH_PATH, paramMap));
+    }
+
+    /**
+     * Calls the GET method for the API which appends the product_id to the path
+     *
+     * @param params VisualSimilarParam specific parameters to perform request
+     *
+     * @return ViSearchHttpResponse http response of search results
+     */
+    public ProductSearchResponse visualSimilarSearch(SearchByIdParam params) {
+        // append the product id after the visual similar path
+        final String path = DEFAULT_VISUAL_SIMILAR_PATH + '/' + params.getProductId();
+        Multimap<String, String> paramMap = params.toMultimap();
+        paramMap.put("app_key", this.appKey);
```

Comment:
missing placementId, need to call same method as imagesearch to inject the placement